### DISCUSSION TITLE
how to use spanbutton  with Self locking button

### DISCUSSION THREAD

**Discussion by ArvinHou** [HomeSpan/HomeSpan#400] [Q&A]
I want to use spanbutton with self hold button,how to do this ?the button
 only have 0=single press (SpanButton::SINGLE)
1=double press (SpanButton::DOUBLE)
2=long press (SpanButton::LONG)
three status.In my case ,is high ,low 2 status. by the way,the self hold button i using is communicate by serial.The code is `newSpanButton(powerPin, Serial_ReadPin);`

**Comment by HomeSpan** [comment]
Though SpanButton will trigger a LONG event once the button is pressed for the specified "long" period of time, it will keep repeating those LONG events after each subsequent "long" period has elapsed.  These subsequent events can be ignored, but SpanButton will unfortunately *not* trigger an event when the button is eventually released, which I think you would need to determine when to power off (assuming the button is controlling some sort of power).

SpanButton is (at present) not well-suitable for this type of button/logic.  However, the good news is you don't have to use SpanButton to read button presses.  SpanButton is more of a convenience function --- its main purpose is to debounce pushbuttons to prevent the generation of dozens of triggers every time a button is pressed or released.  Rather than use SpanButton, you can simply poll the status of a button in the `loop()` method of any Service.  This would be inconvenient if you have to debounce the button (hence the purpose of SpanButton), but in your case I'm guessing your latching button already performs all the debouncing for you since it has a serial interface.

My recommendation would be to poll the status of your button using Serial_ReadPin (or whatever function is needed) in the `loop()` method of the Service you have implemented and simply *compare* the value read with the current status of whatever Characteristic you are trying to control.    If the Characteristic is "on" and your Serial read also returns "on" then do nothing.  Similarly, if they are both "off", do nothing.  If the Serial read returns "on" and the Characteristic is "off", use `setVal()` to turn the Characteristic "on".  And if Serial read returns "off" and the Characteristic is "on", use `setVal()` to turn the Characteristic "off".

Your use-case is quite interesting and I'll see about adding a version of SpanButton() to handle latching in a future release.  Can you provide a link to serial button device?

**Reply by ArvinHou** [reply]
Thank you for your reply.Can you answer the question bellow  ☟

**Comment by ArvinHou** [comment]
Thank you for your reply。The product link is bellow:
https://www.kincony.com/home-automation-local-server.html
https://www.kincony.com/32-channel-ethernet-wifi-relay-module-kc868-h32b.html

**Comment by ArvinHou** [comment]
I want to ask another question.How to use 'setVal()' outside.The example is in header file,in button function.If I want to use set value in ino file.not in class.How to do this? 
such as  the example 15,
![image](https://user-images.githubusercontent.com/11347636/193455942-b48c980c-792e-459f-9e2e-68e205892c89.png)

![image](https://user-images.githubusercontent.com/11347636/193456016-926ecd54-5a5e-42d9-b0c6-92eba4515369.png)

**Comment by HomeSpan** [comment]
There are a number of ways of doing this, but they all rely on storing a pointer to your Service that you can later use in the main `loop()`.  For example, based on the code above, you would do something like this:

```C++
DEV_DimmableLED *myLED;    // create a pointer outside of setup() so it is declared globally

void setup(){
  new SpanAccessory();                                                          
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("PushButton LED");
    
    myLED = new DEV_DimmableLED(17,23,5,18);   // save the pointer to this Service so it can be referenced elsewhere
}

void loop(){
  homeSpan.poll();
  if(a){
  myLED->power->setVal(1-power->getVal());   // use the pointer saved in setup to identify the service
  }
```

Alternatively, you can create a method inside DEV_DimmableLED called something like `togglePower()` that performs whatever logic needed.  Then in the loop() above you would simply call `myLED->togglePower();`

**Reply by ArvinHou** [reply]
Thanks very much!  Now  I know how to do.It would be better if you can update a new spanbutton for self hold button .
